### ISSUE TITLE
Add built-in support for bi-directional deltas in zstd --patch-from

### ISSUE THREAD

**Issue by RubenKelevra** [facebook/zstd#2063]
**Is your feature request related to a problem? Please describe.**

I like to archive older versions of programs like patch does for text-files. But there seems to be no good and fast solution to create delta patches for binary files which can be applied in both directions. 

**Describe the solution you'd like**

The ideal solution would be to specify two uncompressed files and a dictionary that was trained on the filetype and get a small delta that can be applied forwards (to patch the old version to the new one) and backward (to patch the new version to the old one).

This way I can efficiently copy the new file to secondary storage, store it permanently. Then I generate the new version from the old file and delete the old file.

**Describe alternatives you've considered**

There's bsdiff which is quite slow and cannot be trained to be efficient on a specific filetype, and there's courgette that is highly specialized - and cannot be used on random binary files (as far as I understand).

Both are unable to patch in the reverse direction, which is necessary if you want to store the newest file completely not the oldest one. Otherwise, every access to the newest file would require to apply all patches one after another.

**Comment by bimbashrestha** [comment]
Hi @RubenKelevra,

Thanks for submitting the issue. 

We are currently in the process of optimizing zstd for producing patches by introducing a new cli option called --patch-from= which should be a part of the next release! We have been benchmarking our results against bsdiff till now and we are indeed, going to be an order of magnitude faster even with slower zstd levels. We don't have plans for this feature to be bidirectional (ie. you will be able to produce the new file by applying the patch to the old file but you won't be able to produce the old file by applying that SAME patch to the new file). 

Supporting the the reverse operation will be tricky since dictionary compression doesn't intrinsically support that but I think we can flag this as a potential later refinement to the patch-from feature (maybe :)). I'll defer to @Cyan4973 for on whether this is feasible/within the scope of a compression library.

In case you have not already seen the other related thread that sparked the development of this feature (https://github.com/facebook/zstd/issues/1935).

I'd also like to ask you some questions about your use case. How are you planning on using this diff engine? How large are the files you are compressing? If it is possible to share, what files are they? (so far, we've been testing our patch-from on consecutive versions of popular repos on github) Is bidirectionally a core requirement for you?

**Comment by bimbashrestha** [comment]
Also a quick note about this: 


>Both are unable to patch in the reverse direction, which is necessary if you want to store the newest file completely not the oldest one. Otherwise, every access to the newest file would require to apply all patches one after another.


The issue of having to apply consecutive patches starting from the first one can to some extent be mitigated by having patches for consecutive versions as well as patches that skip N versions (so the patch would take version-m to version-(m+N)). This can accelerate patching considerably but it is also slightly more complex. 

This article outlining zstd's usage for patching new releases of the video game league of legends mentions this and some other techniques. Might be worth a look @RubenKelevra
https://technology.riotgames.com/news/supercharging-data-delivery-new-league-patcher

**Comment by bimbashrestha** [comment]
>Will this only work on zstd archives, or is this meant to work on two uncompressed files?

Maybe I misunderstand this question but the usage of the new cli will be: 

```
# create the patch
zstd --patch-from=<old_uncompressed_file> <new_uncompressed_file> -o patch.zst 

# apply the patch
zstd -d --patch-from=<old_uncompressed_file> patch.zst -o <new_uncompressed_file> 
```
>Well, a simple bidirectional version would probably be to concat each of the uncompressed update-commands (forward/backward) and select just the even one for one operation and the uneven ones for the other direction. Regular zstd compression should eliminate the redundancies pretty well without any complex code.

Yes, that would work. I think any approach we take here will involve concatenating two separate patches (one for the forward direction and one for the backward direction). We talked about this in our meeting last week and the verdict we came to was that we would likely not support built-in bi-directionality at least not for the next release because: 
1. We're not convinced that the added benefit of convenience over just having the user of zstd maintain two separate patches on their own is worth the additional complexity..
2. We currently have our hands full for the next release:/

>I'm currently discussing plans to get some compression into IPFS.

Cool! Glad you're considering zstd:) 

Thanks for the detailed report about your use case! Just some overall comments after I read through them: 

Your usage of dictionary compression sounds similar to how zstd is used within Facebook so I suspect that the approach you have laid out will work very well.

As for your segment on delta patching, I see now that your use case is slightly different than the conventional one of incrementally updating existing software. You also plan to use deltas as a way to travel backwards in history to "stored" older versions of files. Unfortunately, like I mentioned, we don't expect built-in bi-directionality to be supported for 1.4.5 (you can of course still use zstd and maintain a concatenated version of a forward and backward patch yourself). 

Perhaps a future release will include this feature? We'll have to gauge how much of a demand exists for it after we release 1.4.5. For the time being, I'm going to mark this feature request as long-term.

**Comment by RubenKelevra** [comment]
> Thanks for the response @RubenKelevra 
> 
> >Will this only work on zstd archives, or is this meant to work on two uncompressed files?
> 
> Maybe I misunderstand this question but the usage of the new cli will be: 
> 
> ```
> # create the patch
> zstd --patch-from=<old_uncompressed_file> <new_uncompressed_file> -o patch.zst 
> 
> # apply the patch
> zstd -d --patch-from=<old_uncompressed_file> patch.zst -o <new_uncompressed_file> 
> ```

You understood the question right.

While beeing able to patch already compressed archives is extremely exiting! Don't get me wrong, the idea was to diff the uncompressed data and compress the patch.

The reasoning is simple: The data might be one blob of binary without a mime type, like scientific raw data. And we like to read and write it on a chunk level, rather than on a file level.

This also allows us to try out every dictionary or a subset of directories on each block, to reach even with files of unknown composition decent compression rates.

This mode of operation (block based access), 
is currently the default.

There's also a block format which stores the block not in a tree but multiple lists, which reduces the overhead for sequencial reading.

This would be optimal if you're expecting the files to be read at once, like a movie.

There are for example the open source movies to try out how IPFS copes with such a kind of data:

https://ipfs.io/ipfs/QmVMntAvXr9AjgGhdGNDbnZR2vZKk38QNh6hZqDLuWgsdp/

---

Bottom line, generating diffs for non-compressed files, but with the help of a dictionary (if we know the mime type) would be pretty cool. Else we would have to add a lot more metadata, to diff each block (and reduce the efficiency massively) - or loose the ability to read files non "as a whole", and store they completely as one large zstd archive.


That said, I'm exited to have the possibility to get fast diffs for Zstd archives, thanks for sharing this neat piece of software with the open source community!




> >Well, a simple bidirectional version would probably be to concat each of the uncompressed update-commands (forward/backward) and select just the even one for one operation and the uneven ones for the other direction. Regular zstd compression should eliminate the redundancies pretty well without any complex code.
> 
> Yes, that would work. I think any approach we take here will involve concatenating two separate patches (one for the forward direction and one for the backward direction). We talked about this in our meeting last week and the verdict we came to was that we would likely not support built-in bi-directionality at least not for the next release because: 
> 1. We're not convinced that the added benefit of convenience over just having the user of zstd maintain two separate patches on their own is worth the additional complexity..
> 2. We currently have our hands full for the next release:/

I fully understand this!

> >I'm currently discussing plans to get some compression into IPFS.
> 
> Cool! Glad you're considering zstd:) 

zstd is an easy choice, in addition to brotli (because of browser support) as a second option.

> Thanks for the detailed report about your use case! Just some overall comments after I read through them: 
> 
> Your usage of dictionary compression sounds similar to how zstd is used within Facebook so I suspect that the approach you have laid out will work very well.

Nice, thanks!

> 
> As for your segment on delta patching, I see now that your use case is slightly different than the conventional one of incrementally updating existing software. You also plan to use deltas as a way to travel backwards in history to "stored" older versions of files. Unfortunately, like I mentioned, we don't expect built-in bi-directionality to be supported for 1.4.5 (you can of course still use zstd and maintain a concatenated version of a forward and backward patch yourself). 

My point was just: When you can concat the two operation - forward and backward - directly behind each other it would be much easier for the compression algorithm to find similar pieces of information, like the addresses of the beginning and end of the patch.

When I store two files, this advantage gets lost and the patches (both together) will get larger.

And yes travelling back in time is actually a bigger goal than having small updates, since bandwidth gets cheaper and cheaper, but storage seems to plateau in the last years pretty much.

Archiving and having a reliability way to get older versions is one of the main cost points apart from having enough money to spend on calculation power and storage - and gets often therefore skipped.

I like to have this feature come to IPFS, to offer easy solution for community based cluster storages which have an reliably, space efficient archiving function right build into it.

So if I had to choose one direction to safe delta patches, it's an easy choice: backwards.

Storing forward patches means, that one patch missing somewhere in between will cut you off from everything newer, while you often just want to jump some versions backwards, not month or years.


> Perhaps a future release will include this feature? We'll have to gauge how much of a demand exists for it after we release 1.4.5. For the time being, I'm going to mark this feature request as long-term.

Thanks!


Best regards,


Ruben

**Comment by RubenKelevra** [comment]
@bimbashrestha 

I was wondering if there's a free library of dictionaries for the different mime types or natural languages etc you're aware of?

AND

If you could recommend a maximal size for a dictionary which is sensible if you put large quantities of data into zstd to generate a mime type dictionary.

Is 1 MB a good choice? Too large? Too small?

**Comment by RubenKelevra** [comment]
I removed my last comment here, I just reread it and it made no sense. I obviously was a bit tired while reading your response. I'm sorry! :)

> Thanks for the response @RubenKelevra 
> 
> >Will this only work on zstd archives, or is this meant to work on two uncompressed files?
> 
> Maybe I misunderstand this question but the usage of the new cli will be: 
> 
> ```
> # create the patch
> zstd --patch-from=<old_uncompressed_file> <new_uncompressed_file> -o patch.zst 
> 
> # apply the patch
> zstd -d --patch-from=<old_uncompressed_file> patch.zst -o <new_uncompressed_file> 
> ```

You understood the question right, thanks - that's actually exactly what I need! :)

> As for your segment on delta patching, I see now that your use case is slightly different than the conventional one of incrementally updating existing software. You also plan to use deltas as a way to travel backward in history to "stored" older versions of files. Unfortunately, as I mentioned, we don't expect built-in bi-directionality to be supported for 1.4.5 (you can of course still use zstd and maintain a concatenated version of a forward and backward patch yourself). 

Indeed, it's easy to store two separate patches for each direction in one archive. But if each operation for a changed segment in a file is saved first forwards and then backward, it's much easier to compress, which will end up a lot of space if there are a lot of versions. :)

---

And yes traveling back in time is actually a bigger goal than having small updates, since bandwidth gets cheaper and cheaper, but storage seems to plateau in the last years pretty much.

Archiving and having a reliable way to get older versions is one of the main cost points apart from having enough money to spend on calculation power and storage - and gets often therefore skipped.

I like to have this feature come to IPFS, to offer an easy solution for community-based cluster storages which have a reliably, space-efficient archiving function right build into it.

So if I had to choose one direction to save delta patches, it's an easy choice: backward.

Storing forward patches means, that one patch missing somewhere in between will cut you off from everything newer, while you often just want to jump some versions backward, not month or years. It also means you have to fully store the oldest version, next to the newest, reducing space efficiency.

And for clearing up old versions, there's the need to apply patches - which will consume a lot of processing power, while if the patches are facing backward, you can just drop a few if you run out of storage space.


> Perhaps a future release will include this feature? We'll have to gauge how much of a demand exists for it after we release 1.4.5. For the time being, I'm going to mark this feature request as long-term.

Thanks for considering!


Best regards

Ruben

**Comment by bimbashrestha** [comment]
>I removed my last comment here, I just reread it and it made no sense. I obviously was a bit tired while reading your response. I'm sorry! :)

No need to apologize. Always appreciate feedback:)

**Comment by RubenKelevra** [comment]
> > I removed my last comment here, I just reread it and it made no sense. I obviously was a bit tired while reading your response. I'm sorry! :)
> 
> No need to apologize. Always appreciate feedback:)

I was wondering if there are free dictionaries for different types of data available you're aware of. I couldn't find one. If not I would go ahead and create some :)

Any recommendations on the settings? Or are the default settings ok when I create a general-purpose dictionary for a type of data?

I currently just plan to increase the size of the dictionary to something like 1 MB.

**Comment by bimbashrestha** [comment]
>I was wondering if there are free dictionaries for different types of data available you're aware of. I couldn't find one. If not I would go ahead and create some :)

To my knowledge, there is not. I can see this being a very valuable thing to do though:)

>Any recommendations on the settings? Or are the default settings ok when I create a general-purpose dictionary for a type of data?

I think the default dictionary trainer should do a good job. If you do go ahead and create this, it might also be a good idea to periodically (automatically maybe?) train with new data of each type and replace the existing dictionary if the new trained dictionary performs better on some sort of separate validation set. @felixhandte might be able to chip in here since he's done a lot of this type of work internally for us.

>I currently just plan to increase the size of the dictionary to something like 1 MB.

Playing around with dictionary sizes is a good idea. Athough I believe for most data types, the same compression ratio can be obtained using smaller dictionaries (110K which is the default for example).

**Comment by RubenKelevra** [comment]
> Playing around with dictionary sizes is a good idea. Athough I believe for most data types, the same compression ratio can be obtained using smaller dictionaries (110K which is the default for example).

I tried 1 MB and the default, for the Linux Kernel corpus. With compression level 22 the 1 MB dict was beneficial, while levels 3, 9 and 15 were actually hurt by the larger dictionary - quite unexpected.

I created a dictionary explicitly for the compression level on which I used it afterward - so it was somewhat strange that a larger dictionary would hurt it. 🤔

The funniest thing was actually, to look into the dictionary. There are extremely large snippets of source code. Sure they might be copied and pasted in multiple source files by the developers, or the dict generator just copied large snippets in the database to avoid having to save them in the archive :laughing: 

The Linux archive dict:
https://ipfs.io/ipfs/bafybeiblnqhkxiicwftnihtvrwwzzd5ba6g3ieejng5s2b5hkn3zhtqafi

> I think the default dictionary trainer should do a good job. If you do go ahead and create this, it might also be a good idea to periodically (automatically maybe?) train with new data of each type and replace the existing dictionary if the new trained dictionary performs better on some sort of separate validation set. @felixhandte might be able to chip in here since he's done a lot of this type of work internally for us.

That's a good idea, but a thing I won't do.

The idea was to use small byte long integers to address those dictionaries and match them up to a list stored in the application. So if the user wants to store some say SQL files, the pretrained dictionary can be fetched from the list and be used on those files.

Having different versions means we have to use larger integers for the dictionary and may have blocks from the same format compressed with different dictionaries, which means we have to hold multiple different versions on a node.

This can get pretty messy and hard to debug tbh.



> Quick additional note:
> 
> Dictionaries are most useful when the data you're trying to optimize for has high intra-sample redundancy. So in other words, the more "narrow" your data set, the better. Having general purpose dictionaries might work if the "general" case is very similar in content (but then it might not be so general anymore).

Well, I'm currently working on a dictionary for English, which should optimize itself down (in my theory) to the most used terms. I thought to push some email archives and some Wikipedia articles into the algorithm and look how the dictionary performs after that. :)

**Comment by bimbashrestha** [comment]
Hey @RubenKelevra, 

Just wanted to inform you that we've merged --long range support for the patch-from feature into zstd dev today. Here is a brief summary: https://github.com/facebook/zstd/pull/1959#issue-364364610

Unless other developments arise, this is what will make it to the release. Would love it if you gave it a shot on your use case at some point. Hopefully the long mode should provide a nice boost to compression ratio from before. 

Bimba

**Comment by RubenKelevra** [comment]
> Just wanted to inform you that we've merged --long range support for the patch-from feature into zstd dev today. Here is a brief summary: https://github.com/facebook/zstd/pull/1959#issue-364364610
> 
> Unless other developments arise, this is what will make it to the release. Would love it if you gave it a shot on your use case at some point. Hopefully the long mode should provide a nice boost to compression ratio from before. 
> 
> 

Hey Bimba,

thanks for the heads-up, sounds great!

I will check that on some sample data and report back, thanks!

--

Is there any documentation about your delta patch algorithm?

I haven't looked at the code yet... but I gave the problem some thoughts. Guess that a patch format for binaries would look something like this:

* offset in bits
* old length in bits
* drop flag (beginning or end)
* patch mode (replace/xor)
* patch

Since you are using the dictionary function for that I suspect that your approach is more like a 1:1 search pattern to be replaced by new data.

This means each new patch needs to make sure it's pattern is either unique or the replacement is everywhere the same, while on smaller patterns a direct comparison is done, on larger ones a hashsum is used.

 

How far off am I? :)


Best regards

Ruben

**Comment by bimbashrestha** [comment]
Hi @RubenKelevra,

> Since you are using the dictionary function for that I suspect that your approach is more like a 1:1 search pattern to be replaced by new data.

If I understand this correctly, this is more akin to what is happening within patch-from.

There isn't a new algorithm introduced exactly. We've simply extended the existing LZ style match finding (which is primarily what sets different compression levels apart) a little bit to allow for better results when dealing with larger dictionaries. 

>I haven't looked at the code yet... but I gave the problem some thoughts. Guess that a patch format for binaries would look something like this:
>
>offset in bits
>old length in bits
>drop flag (beginning or end)
>patch mode (replace/xor)
>patch

We are finding exact matches between the 'source' (which in this case is the new version of the file) and the 'dictionary' (the old version) and encoding them as before into the [zstd compression format](https://github.com/facebook/zstd/blob/dev/doc/zstd_compression_format.md). The details of the representation are not unique to the patch-from feature.

Hope that answered your question. I'm happy to go into more detail about the actual match finding algorithms we have if you'd like but those are generic to regular (non patch-from) compression.

Bimba

**Comment by RubenKelevra** [comment]
Thanks!

I'm was wondering if there's a chance to speed things up even more:

IPFS has the ability to dedup blocks between different types of files, for this they use a rolling hash algorithm.

You can either select rabin or buzzhash for this task. Rabin is kind of slow, but buzzhash is quite fast.

The rolling hash would allow to 'prescan' both files, get some cut marks and run some fast cryptographic hash algorithm over the chunks, like blake2b.

I think both operations are much cheaper than pattern matching. This way you can skip all pattern matching attempts which are on both sides (A and B) inside the known equal blocks.

The first layer of patching would just generate a lengths+offset+move triple, which can copy the blocks from the original file into a sparse file as first patching operation.

The pattern matching rules would be used on top of that, completing the gaps of the output file.

**Comment by bimbashrestha** [comment]
Hi @RubenKelevra sorry for the super late response. 

Yes! there are definitely ways to improve speed and ratio here. But implementing those will involve more discussion. I'm going to close this particular issue for now as the feature has been merged. 

Let's this discussion on https://github.com/facebook/zstd/issues/2173 which also requests an extension to --patch-from. Our issues are getting a bit cluttered so I'm going to consolidate:)

**Comment by RubenKelevra** [comment]
Yeah it got pretty off-topic, sorry for that. I've created a new ticket for that. :)
